File: dfs.py
from graph import Graph

class DepthFirstSearch:
    def __init__(self, g, s):
        self.s = s
        self.marked = {}
        self.edgeTo = {}
        self.__dfs(g, s)

    def hasPathTo(self, v):
        return v in self.marked

    def pathTo(self, v):
        path = []
        while v != self.s:
            path.insert(0, v)
            v = self.edgeTo[v]
        path.insert(0, self.s)
        return path

    def __dfs(self, g, s):
        self.marked[s] = True
        for w in g.getAdj(s):
            if w not in self.marked:
                self.edgeTo[w] = s
                self.__dfs(g, w)

if __name__ == "__main__":

    g = Graph("exemplos/tinyG.txt")

    dfs = DepthFirstSearch(g, "0")

    for v in g.getVerts():
        print(f"{v}: ", end="")
        if dfs.hasPathTo(v):
            for w in dfs.pathTo(v):
                print(f"{w} ", end="")
            print()
        else:
            print("SEM CAMINHO")
    print()

File: DFS.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Map;
import java.util.Set;

public class DFS {
    private Set<String> marked;
    private Map<String, String> edgeTo;
    private String start;

    public DFS(Graph g, String start) {
        this.start = start;
        marked = new HashSet<>();
        edgeTo = new HashMap<>();
        edgeTo.put(start, null);
        dfs(g, start);
    }

    public boolean hasPathTo(String v) {
        return edgeTo.containsKey(v);
        // return marked.contains(v);
    }

    public Iterable<String> pathTo(String v) {
        LinkedList<String> path = new LinkedList<>();
        while (v != null) {
            path.add(0, v);
            v = edgeTo.get(v);
        }
        return path;
    }

    public void dfs(Graph g, String v) {
        System.out.println("Visitando: " + v);
        marked.add(v);
        for (String w : g.getAdj(v)) {
            if (!marked.contains(w)) {
                edgeTo.put(w, v);
                dfs(g, w);
            }
        }
    }
}